### ISSUE TITLE
vmagent: periodically panic with missed metainfo.json

### ISSUE THREAD

**Issue by vandud** [VictoriaMetrics/VictoriaMetrics#5357]
### Describe the bug

vmagents periodically get panic with this error:
```
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: 2023-11-20T15:30:08.511Z        panic        VictoriaMetrics/lib/persistentqueue/persistentqueue.go:386        FATAL: cannot flush metainfo: cannot write metainfo to "/var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json": cannot write persistent queue metainfo to "/var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json": open /var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json: no such file or directory
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: panic: FATAL: cannot flush metainfo: cannot write metainfo to "/var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json": cannot write persistent queue metainfo to "/var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json": open /var/lib/vmagent/persistent-queue/1_1458901A2D90C8F2/metainfo.json: no such file or directory
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: goroutine 3263 [running]:
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logMessage({0xa4338b, 0x5}, {0xc00011a9a0, 0x14f}, 0x1?)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:301 +0xa91
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevelSkipframes(0x1, {0xa4338b, 0x5}, {0xa44fce?, 0xc013830000?}, {0xc00221bcc0?, 0x40770b?, 0xc00cdd8000?})
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:141 +0x1a5
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.logLevel(...)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:133
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/logger.Panicf(...)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/logger/logger.go:129
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue.(*queue).MustWriteBlock(0xc0000e0680, {0xc009924000, 0x139, 0x7000})
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue/persistentqueue.go:386 +0x285
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue.(*FastQueue).flushInmemoryBlocksToFileLocked(0xc0052645b0)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue/fastqueue.go:108 +0x5c
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue.(*FastQueue).flushInmemoryBlocksToFileIfNeededLocked(...)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue/fastqueue.go:101
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue.(*FastQueue).MustWriteBlock(0xc0052645b0, {0xc00cdd8000, 0x12f, 0x10000})
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/lib/persistentqueue/fastqueue.go:140 +0xed
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.pushWriteRequest(0xc003310b70, 0xc0067168d0, 0x1)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/pendingseries.go:215 +0x267
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.(*writeRequest).flush(0xc003310b48)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/pendingseries.go:137 +0x6e
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.(*pendingSeries).periodicFlusher(0xc003310b40)
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/pendingseries.go:84 +0x1be
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.newPendingSeries.func1()
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/pendingseries.go:50 +0x50
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]: created by github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.newPendingSeries in goroutine 101
Nov 20 18:30:08 mon-vm-01 vmagent-prod[26574]:         github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/pendingseries.go:48 +0x105
Nov 20 18:30:08 mon-vm-01 systemd[1]: vmagent.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Nov 20 18:30:08 mon-vm-01 systemd[1]: vmagent.service: Failed with result 'exit-code'.
```

I dont know why this happens
I tried to increase the limit on the number of open files, but it had no effect

### To Reproduce

This happens several times per hour 
Error message is above

systemd unit:
```
[Unit]
Description=vmagent
After=network.target

[Service]
Type=simple
User=vmagent
Group=vmagent
StartLimitBurst=5
StartLimitInterval=5
Restart=always
RestartSec=5
EnvironmentFile=/etc/vmagent/vmagent.conf
ExecStart=/usr/local/bin/vmagent-prod -envflag.enable
ExecReload=/bin/kill -s SIGHUP $MAINPID
ExecStop=/bin/kill -s SIGTERM $MAINPID

LimitNOFILE=10240

[Install]
WantedBy=multi-user.target
```
vmagent.conf with env vars:
```
promscrape_cluster_memberNum=0
promscrape_cluster_membersCount=5
promscrape_cluster_replicationFactor=2
promscrape_config=/etc/vmagent/main.yaml
remoteWrite_multitenantURL=http://mon-vm-01:8480,http://mon-vm-02:8480,http://mon-vm-03:8480,http://mon-vm-04:8480,http://mon-vm-05:8480
remoteWrite_tmpDataPath=/var/lib/vmagent
remoteWrite_maxDiskUsagePerURL=2GB
remoteWrite_showURL=True
remoteWrite_relabelConfig=/etc/vmagent/global-relabeling.yaml
promscrape_suppressScrapeErrors=True
promscrape_suppressScrapeErrorsDelay=10m
maxConcurrentInserts=32
memory_allowedBytes=2GB
```

query `sum(vm_promscrape_targets)` returning 1664 targets

### Version

vmagent-20231116-193721-tags-v1.95.1-0-g354563393

**Comment by vandud** [comment]
@zekker6 
/var/lib/vmagent/persistent-queue/ - ext4 (server is Debian 10 virtual machine)
I can try to run vmagent in docker container without volume mapping if it really need to, but it can spend some time
Pattern in time is indistinguishable

**Comment by zekker6** [comment]
@vandud 
>  I can try to run vmagent in docker container without volume mapping if it really need to, but it can spend some time

This can be useful to narrow down the scope of things that affect `vmagent` disk.

I've also set up a `vmagent` in a debian VM for testing to see if it will experience the same issue as you've posted.

**Comment by YuriPachkovsky** [comment]
I use vmagent in a container and checked container + ephemeral, container + rbd, container + hostpat - the error is reproduced in my environment. The frequency is chaotic, but if there are problems in network connectivity with vmstorage, the error frequency increases.

**Comment by f41gh7** [comment]
@zekker6  I suspect, that there is a race condition. I recommend to investigate behavior of `(q *queue) readBlock` function, it can be called with blocked `MustWriteBlock` and non blocked `MustReadBlockNonblocking` modes.

 This function may execute `skipBrokenChunkFile` method that deletes directory content via `mustResetFiles` call.

 My assumption is, if some process modifies chunk files on disk and corrupts those files. It may lead to race condition and panic.

**Comment by zekker6** [comment]
@f41gh7 Thanks for the idea!

@YuriPachkovsky @vandud Could you also check if there are any error logs messages before the panic happens?
As per suggestion above, it is possible that vmagent will drop directory content in case it will detect invalid content and this should trigger writing error log line.

**Comment by valyala** [comment]
@vandud , could you check `vmagent` logs for messages containing the `skipping corrupted` text and post these messages here? Could you also post the last messages in `vmagent` log just before the panic occurs?

**Comment by valyala** [comment]
The issue has been identified - it is triggered when [`-remoteWrite.multitenantURL`](https://docs.victoriametrics.com/vmagent.html#multitenancy) command-line flag is set and when samples for multiple tenants are simultaneously written to remote storage. The root cause of the issue is incorrect deletion of persistent queues when new queues are created for the new tenant. The issue has been fixed in 1831c731a3441baf1b2d26eb3370dd8c9017eb39 .

BTW, @vandud , why you specify multiple urls at `-remoteWrite.multitenantURL` command-line flag?

```
remoteWrite_multitenantURL=http://mon-vm-01:8480,http://mon-vm-02:8480,http://mon-vm-03:8480,http://mon-vm-04:8480,http://mon-vm-05:8480
```

`vmagent` writes the same data simultaneously (e.g. it replicates this data) to all the specified `-remoteWrite.multitenantURL` remote storage systems. It is likely this is not what you want. If you spreading the data among multiple `vminsert` nodes in a single [VictoriaMetrics cluster](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html), then it is recommended putting [vmauth](https://docs.victoriametrics.com/vmauth.html) in front of these `vminsert` nodes as described [here](https://docs.victoriametrics.com/vmauth.html#load-balancing).

**Comment by vandud** [comment]
Sorry for the long wait
If it's still useful
My vmagent logs not contains the lines with `corrupted` pattern
There is vmagent log from start to panic - https://pastebin.com/BnAmtGd6

Thanks for comment about `-remoteWrite.multitenantURL`
Before I used loadbalancing nginx in front of vminserts, but I thought it is dangerous to have a single point of failure and I setup the direct connection + I wanted to exclude balancing as error cause

**Comment by valyala** [comment]
FYI, the bugfix has been included into [v1.93.9 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.9). It will be included also into the latest upcoming release.
